routes/dao_coin_exchange.go

```diff
@@ -0,0 +1,183 @@
+package routes
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/deso-protocol/core/lib"
+	"github.com/holiman/uint256"
+	"io"
+	"math/big"
+	"net/http"
+)
+
+type GetDAOCoinLimitOrdersRequest struct {
+	DAOCoin1CreatorPublicKeyBase58CheckOrUsername string `safeForLogging:"true"`
+	DAOCoin2CreatorPublicKeyBase58CheckOrUsername string `safeForLogging:"true"`
+}
+
+type GetDAOCoinLimitOrdersResponse struct {
+	Orders []DAOCoinLimitOrder
+}
+
+type DAOCoinLimitOrder struct {
+	TransactorPublicKeyBase58CheckOrUsername string `safeForLogging:"true"`
+
+	BuyingDAOCoinCreatorPublicKeyBase58CheckOrUsername  string `safeForLogging:"true"`
+	SellingDAOCoinCreatorPublicKeyBase58CheckOrUsername string `safeForLogging:"true"`
+
+	// One of these two should be populated
+	ScaledExchangeRateCoinsToSellPerCoinToBuy *uint256.Int `safeForLogging:"true"`
+	ExchangeRateCoinsToSellPerCoinToBuy       float64      `safeForLogging:"true"`
+
+	// One of these two should be populated
+	QuantityToBuyInBaseUnits *uint256.Int `safeForLogging:"true"`
+	QuantityToBuy            float64      `safeForLogging:"true"`
+}
+
+func (fes *APIServer) GetDAOCoinLimitOrders(ww http.ResponseWriter, req *http.Request) {
+	decoder := json.NewDecoder(io.LimitReader(req.Body, MaxRequestBodySizeBytes))
+	requestData := GetDAOCoinLimitOrdersRequest{}
+	if err := decoder.Decode(&requestData); err != nil {
+		_AddBadRequestError(
+			ww,
+			fmt.Sprintf("GetDAOCoinLimitOrders: Problem parsing request body: %v", err),
+		)
+		return
+	}
+
+	if requestData.DAOCoin1CreatorPublicKeyBase58CheckOrUsername == "" &&
+		requestData.DAOCoin2CreatorPublicKeyBase58CheckOrUsername == "" {
+		_AddBadRequestError(
+			ww,
+			fmt.Sprint("GetDAOCoinLimitOrders: Must provide either a "+
+				"DAOCoin1CreatorPublicKeyBase58CheckOrUsername or DAOCoin2CreatorPublicKeyBase58CheckOrUsername "+
+				"or both"),
+		)
+		return
+	}
+
+	utxoView, err := fes.backendServer.GetMempool().GetAugmentedUniversalView()
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("GetDAOCoinLimitOrders: Problem fetching utxoView: %v", err))
+		return
+	}
+
+	coin1PKID := &lib.ZeroPKID
+	coin2PKID := &lib.ZeroPKID
+
+	coin1ProfilePublicKeyBytes := lib.ZeroPublicKey.ToBytes()
+	coin2ProfilePublicKeyBytes := lib.ZeroPublicKey.ToBytes()
+
+	coin1ProfilePublicBase58Check := ""
+	coin2ProfilePublicBase58Check := ""
+
+	if requestData.DAOCoin1CreatorPublicKeyBase58CheckOrUsername != "" {
+		coin1ProfilePublicKeyBytes, _, err = fes.GetPubKeAndProfileEntryForUsernameOrPublicKeyBase58Check(
+			requestData.DAOCoin1CreatorPublicKeyBase58CheckOrUsername,
+			utxoView,
+		)
+		if err != nil {
+			_AddBadRequestError(
+				ww,
+				fmt.Sprint("GetDAOCoinLimitOrders: Invalid DAOCoin1CreatorPublicKeyBase58CheckOrUsername"),
+			)
+			return
+		}
+		coin1PKID = utxoView.GetPKIDForPublicKey(coin1ProfilePublicKeyBytes).PKID
+		coin1ProfilePublicBase58Check = lib.Base58CheckEncode(coin1ProfilePublicKeyBytes, false, fes.Params)
```

Comment:
can probably refactor this into a function since this is all duplicated below. It can be a function that's defined within this function if that makes some things easier.